### PR TITLE
[MOD] BART TASK Update

### PR DESCRIPTION
**Pull request by hmy831004**
## Title
BART TASK SOURCE UPDATE 

## Description
BART DENOSING SOURCE Modifying, 
BART Task Test source

### CODE REVIEW

**Comment by hyunwoongko** [comment]
@hmy831004 thanks for great PR. could we review this code together tomorrow?

**Comment by hmy831004** [comment]
Okay, I'll review it today.

File: oslo/transformers/tasks/data_bart_pretraining.py [hunk 1]
```diff
@@ -1,92 +1,132 @@
...
-                chunk_ids = self._buffer[: self._chunk_size]
-                dict_of_training_examples["input_ids"].append(chunk_ids)
-                self._buffer = self._buffer[self._chunk_size :]
-
-        return dict_of_training_examples
+        
+        def split_sentence_add_token(example):
+            # l reprise the role in the last two films..  Watch I-Reporter give her review of Potter's latest » . There is life beyond 
+            # pattern = r"(?<=\w[.!?])(?<![A-Z][.!?])(?<![0-9][.!?])|(?<=\s[.!?])"
+            pattern = r"(?<=\w[.!?])(?<![A-Z][.!?])(?<![0-9][.!?])"
```
Comment (hmy831004) [inline oslo/transformers/tasks/data_bart_pretraining.py:41]: add sentence split pattern

File: oslo/transformers/tasks/data_bart_pretraining.py [hunk 2]
```diff
@@ -1,92 +1,132 @@
...
+            splited_sents = [
```
Comment (hmy831004) [inline oslo/transformers/tasks/data_bart_pretraining.py:42]: After separating sentences, spatial_token add

File: oslo/transformers/tasks/data_bart_pretraining.py [hunk 3]
```diff
@@ -1,92 +1,132 @@
...
+            self._buffer.extend(list(chain.from_iterable(examples['input_ids']))) # combine input_ids dataset 
+            total_length = len(self._buffer)
+            quotient, remain = divmod(total_length, self._max_seq_length)                            
+            output = [ self._buffer[i : i + self._max_seq_length] 
+                    for i in range(0, quotient * self._max_seq_length, self._max_seq_length)
+                    ] 
+            self._buffer = self._buffer[quotient * self._max_seq_length:]
+            return {'input_ids':output}
+        
+        group_dataset = split_tokenized_dataset.map(
```
Comment (hmy831004) [inline oslo/transformers/tasks/data_bart_pretraining.py:83]: Group separated sentences by max_seq_length.
After grouping, the rest of the tokens are put in the buffer so that nothing is thrown away.

File: oslo/transformers/tasks/data_bart_pretraining.py [hunk 4]
```diff
@@ -102,122 +142,149 @@ def __init__(
...
-        # Shift input ids one token to the right
-        shifted_input_ids = batch["labels"].new_zeros(batch["labels"].shape)
-        shifted_input_ids[:, 1:] = batch["labels"][:, :-1].clone()
-        shifted_input_ids[:, 0] = self.decoder_start_token_id
-
-        shifted_input_ids.masked_fill_(
-            shifted_input_ids == self.label_pad_token_id,
-            self.pad_token_id,
-        )
+    def permute_sentences(self, input_ids):
```
Comment (hmy831004) [inline oslo/transformers/tasks/data_bart_pretraining.py:168]: permute sentence function - 

Check the end point of the sentence based on pad_token and mix the sentences according to the percentage of permit_sentence_ratio.

File: oslo/transformers/tasks/data_bart_pretraining.py [hunk 5]
```diff
@@ -102,122 +142,149 @@ def __init__(
...
+            # write shuffled sentences into results
+            # Need to solve the problem of having to repeat the whole sentence, later.
+            index = 0
+            for j in ordering:
+                sentence = input_ids[i, (sentence_ends_map[i][j - 1] if j > 0 else 0) : sentence_ends_map[i][j]]
+                results[i, index : index + sentence.shape[0]] = sentence
+                index += sentence.shape[0]
+        return results  
+
+    def text_infilling(self, input_ids, labels, do_permute):
```
Comment (hmy831004) [inline oslo/transformers/tasks/data_bart_pretraining.py:211]: text_infilling function - 

We randomly set tokens in sentences and mask them.
Determined how many tokens you want to extend according to the Poisson distribution.

I don't make special_token or pad_token masking.

File: oslo/transformers/tasks/data_bart_pretraining.py [hunk 6]
```diff
@@ -102,122 +142,149 @@ def __init__(
...
+            # Need to modify only the mainnig part repeatedly without repeating all masked_indices, B.S
+            masked_indices[remaining, 1] += 1
+            for mi in masked_indices:
+                mask[tuple(mi)] = True
+            span_lengths -= 1
+            remaining = (span_lengths > 0) & (masked_indices[:, 1] < max_index)
+
+        # place the mask tokens
+        # If the mask corresponds to special_tokens_mask_inputs, enter False because you should not mask
+        mask[np.where(special_tokens_mask_inputs)] = False
```
Comment (hmy831004) [inline oslo/transformers/tasks/data_bart_pretraining.py:275]: Issue
Line 277 ~ 280 why need?
I want to talk about this.

Should the special token always be removed from the label?

File: oslo/transformers/tasks/data_bart_pretraining.py (hunk 1 above)
Comment (hyunwoongko) [reply oslo/transformers/tasks/data_bart_pretraining.py:41]: `(?<=[.!?]\s)`

Comment (hyunwoongko) [reply oslo/transformers/tasks/data_bart_pretraining.py:41]: `re.split(r'(?<=[.!?])\s', text)`

File: oslo/transformers/tasks/data_bart_pretraining.py [hunk 7]
```diff
@@ -1,92 +1,132 @@
...
-
-        return dict_of_training_examples
+        
+        def split_sentence_add_token(example):
+            # l reprise the role in the last two films..  Watch I-Reporter give her review of Potter's latest » . There is life beyond 
+            # pattern = r"(?<=\w[.!?])(?<![A-Z][.!?])(?<![0-9][.!?])|(?<=\s[.!?])"
+            pattern = r"(?<=\w[.!?])(?<![A-Z][.!?])(?<![0-9][.!?])"
+            splited_sents = [
+                self._tokenizer.bos_token 
+                + f'{self._tokenizer.pad_token}'.join(
```
Comment (hyunwoongko) [inline oslo/transformers/tasks/data_bart_pretraining.py:44]: self._tokenizer.sep_token

Comment (hyunwoongko) [reply oslo/transformers/tasks/data_bart_pretraining.py:44]: `<bos>문장1<sep>문장2<sep>.....<eos>`

File: oslo/transformers/tasks/data_bart_pretraining.py [hunk 8]
```diff
@@ -1,92 +1,132 @@
...
+                    return_special_tokens_mask=False,
+                    verbose=False,
+            )['input_ids']
+            
+            return {'input_ids':tokenized_sents}
+        
+        split_tokenized_dataset =  dataset.map(
+            split_sentence_add_token,
+            batched = True,
+            num_proc = 2, #  number of multiprocessor
```
Comment (hyunwoongko) [inline oslo/transformers/tasks/data_bart_pretraining.py:69]:
```python
import torch.distributed as dist
import os

if dist.is_initialized():
    world_size = dist.get_world_size()
else:
    world_size = 1

split_tokenized_dataset =  dataset.map(
    split_sentence_add_token,
    batched = True,
    num_proc = os.cpu_count() // world_size,
    remove_columns=column_names 
)
```

File: oslo/transformers/tasks/data_bart_pretraining.py (hunk 3 above)
Comment (hyunwoongko) [reply oslo/transformers/tasks/data_bart_pretraining.py:83]: sample1: `<bos>문장1<sep>문장2<sep>문장3` <-- no eos
sample2: `<sep>....<sep>문장10000<eos>` <-- no bos

File: oslo/transformers/tasks/data_bart_pretraining.py [hunk 9]
```diff
@@ -1,92 +1,132 @@
...
+            output = [ self._buffer[i : i + self._max_seq_length] 
+                    for i in range(0, quotient * self._max_seq_length, self._max_seq_length)
+                    ] 
+            self._buffer = self._buffer[quotient * self._max_seq_length:]
+            return {'input_ids':output}
+        
+        group_dataset = split_tokenized_dataset.map(
+            group_text,
+            batched=True,
+            num_proc=2
```
Comment (hyunwoongko) [inline oslo/transformers/tasks/data_bart_pretraining.py:86]: num_proc -> os.cpu_count() // world_size

File: oslo/transformers/tasks/data_bart_pretraining.py [hunk 10]
```diff
@@ -1,92 +1,132 @@
...
     """
 
     def __init__(
         self,
         processor: ProcessorForBartPretraining,
-        mlm_probability: float = 0.15,
-        possion_lambda: float = 3.0,
-        permute_sentence: bool = True,
-        label_pad_token_id: int = -100,
+        mask_ratio: float = 0.3,
```
Comment (hyunwoongko) [inline oslo/transformers/tasks/data_bart_pretraining.py:114]: variable name is needed to check

File: oslo/transformers/tasks/data_bart_pretraining.py [hunk 11]
```diff
@@ -1,92 +1,132 @@
...
+        poisson_lambda: float = 3.0,
+        permute_sentence_ratio: bool = 0.5,
         decoder_start_token_id: Optional[int] = None,
+        label_pad_token_id: int = -100,
+        parallel_context: Optional[ParallelContext] = None
```
Comment (hyunwoongko) [inline oslo/transformers/tasks/data_bart_pretraining.py:119]: 사용하지 않는 코드 -> 제거하기.

**Comment by hyunwoongko** [comment]
`pip install pre-commit`
`precommit run --all-files`

File: oslo/transformers/tasks/data_bart_pretraining.py (hunk 10 above)
Comment (hmy831004) [reply oslo/transformers/tasks/data_bart_pretraining.py:114]: These words (mask_ratio, permute_sentence_ratio, max_seq_length) are used in the huggingface.

File: oslo/transformers/tasks/data_bart_pretraining.py (hunk 11 above)
Comment (hmy831004) [reply oslo/transformers/tasks/data_bart_pretraining.py:119]: done.
